### ISSUE TITLE
TF2 Warning :  Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.

### ISSUE THREAD

**Issue by songs18** [tensorflow/tensorflow#34369]
When i compute grads by tf.GradientTape, console presents  a **warning : Sets are not currently considered sequences, but this may change in future, so consider avoiding using them.** My code is like this 
```python
import tensorflow as tf
import numpy as np
import logging

logging.basicConfig(format='%(message)s', level=logging.INFO)

input = np.array([[0.1, 0], [0, 0.1]], np.float32)
target = np.array([1, 0], np.float32)


class layer(tf.Module):
    def __init__(self):
        super(layer, self).__init__()
        with self.name_scope:
            self.w1 = tf.Variable(tf.ones([2, 2]), name='w1')

    @tf.Module.with_name_scope
    def __call__(self, x):
        return tf.nn.tanh(x * self.w1)


def get_loss(pred, y_):
    return tf.losses.binary_crossentropy(pred, y_)


def get_opti():
    return tf.optimizers.SGD()


my_layer = layer()

optimizer = get_opti()

for _ in range(2):
    for x_, y_ in zip(input, target):
        with tf.GradientTape() as tape:
            pred = my_layer(x_)

            loss = get_loss(pred, y_)
        grads = tape.gradient(loss, my_layer.trainable_variables )
        optimizer.apply_gradients(zip(grads, my_layer.trainable_variables ))


print('-----------')
print(my_layer.w1.numpy())
```
![image](https://user-images.githubusercontent.com/47944452/69391681-8a58cc80-0d0e-11ea-9b73-f35742116387.png)

**Comment by gadagashwini** [comment]
@songs18, Please provide the complete code and the Tensorflow version. Thanks!

**Comment by songs18** [comment]
Duplicate of #
The code is very simple, it contains just single hidden layer. In the following code, train_x refers input x, ground_truth is the correct label. I am doing a classification task. 

` with tf.GradientTape() as tape:
            pred,w = model(train_x)
            loss_pred = compute_loss(ground_truth, pred)

        grads = tape.gradient(loss_pred, model.trainable_variables)

         optimizer.apply_gradients(zip(grads, model.trainable_variables))`

The version of tensorflow  is v2.0.0.
Note: When i use v2.0.0 release version, there is no warning, but i update it to the lastest version, it appears the W:
2019-11-19 15:55:58.350407: W tensorflow/python/util/util.cc:299] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
I looked the util.cc in tensorflow source code, it is likely the parameter of the tape.gradient() input could change in future. But i don't know how to write the compute gradient code is the best way.
Thanks!

**Comment by gadagashwini** [comment]
@songs18, Just to verify, Did you update the latest version using 
`pip install -U tensorflow` ?

**Comment by songs18** [comment]
No, the environment i used is a sandbox created by virtualenv, i installed tensorlfow 2.0.0 just by insert the command 'pip install tensorflow'

**Comment by songs18** [comment]
sorry, my tensorflow is actually the lastest version, i tried -U, the warning remains.

**Comment by gadagashwini** [comment]
@songs18, Can you add the complete code which generates this warning? It's hard to say whether it's important without knowing what you were trying to do. It will indeed help us to move faster. Thanks!

**Comment by songs18** [comment]
To run any tf2 script will reproduce the warning, so i do not provide the whole code but just the exact code snippet, if you need the whole code, i can provide a simple verison.
`import tensorflow as tf
import numpy as np
import logging

logging.basicConfig(format='%(message)s', level=logging.INFO)

input = np.array([[0.1, 0], [0, 0.1]], np.float32)
target = np.array([1, 0], np.float32)


class layer(tf.Module):
    def __init__(self):
        super(layer, self).__init__()
        with self.name_scope:
            self.w1 = tf.Variable(tf.ones([2, 2]), name='w1')

    @tf.Module.with_name_scope
    def __call__(self, x):
        return tf.nn.tanh(x * self.w1)


def get_loss(pred, y_):
    return tf.losses.binary_crossentropy(pred, y_)


def get_opti():
    return tf.optimizers.SGD()


my_layer = layer()

optimizer = get_opti()

for _ in range(2):
    for x_, y_ in zip(input, target):
        with tf.GradientTape() as tape:
            pred = my_layer(x_)

            loss = get_loss(pred, y_)
        grads = tape.gradient(loss, my_layer.trainable_variables )
        optimizer.apply_gradients(zip(grads, my_layer.trainable_variables ))


print('-----------')
print(my_layer.w1.numpy())
`
![image](https://user-images.githubusercontent.com/47944452/69391681-8a58cc80-0d0e-11ea-9b73-f35742116387.png)

**Comment by songs18** [comment]
From the warning info, the code can work now, but cannot promise in future, i want to write right code which meet the tf recommends, unfortunately, i do not get help from tf website, i want to know how to write code to eliminate the warning.
Thanks!

**Comment by gadagashwini** [comment]
@songs18, Thanks for the code, I tried replicating the reported issue but looks like code is not intended correctly. Please take a look at [gist](https://colab.sandbox.google.com/gist/gadagashwini/08de511173260cc94e849e9a0ea1e5bc/untitled273.ipynb) and modify accordingly. Thanks!

**Comment by songs18** [comment]
Thanks for your patient, unfortunately, i can not open the gist link as i am in China and i cannot use Gloole any service, i post my snippet in [Gitee](unfortunately) which is one webiset like github in China.

**Comment by gadagashwini** [comment]
@songs18, The warnings are about Tensorflow support for AVX instruction sets. 
```
TensorFlow release binaries version 1.6 and higher are prebuilt with AVX instruction sets.
Therefore on any CPU that does not have these instruction sets, either CPU or GPU version of TF will fail to load.
``` 
For more read [here](https://www.tensorflow.org/install/pip?lang=python3#hardware-requirements). Your code is good to go, warnings can be ignored. Thanks!

**Comment by songs18** [comment]
Thanks. I read the page, unfortunately, i do not understand how AVX causes the warning, could you provide a detailed explain? What is the relationship between AVX and the warning?

**Comment by gadagashwini** [comment]
@songs18, If you use pip version of tensorflow, which means it's already compiled and you are just installing it. When you download it from repository and trying to build, you should build it with CPU AVX support. If you ignore it, you will get the warning every time when you run on cpu.
If you don't want to get these warnings, You can turn it off by setting os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'.

**Comment by songs18** [comment]
Thanks.
I know that AVX could cause the warning ' Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA',  and my question is not this one.
'Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them'  occured during computing derivative and applying gradient, how AVX cuase this warning? Is there any relationship between AVX and computing derivative and applying gradient? 
Thanks for your long time patient and help !!

**Comment by 306404391** [comment]
@songs18 解决了吗?

**Comment by cypherix** [comment]
@ymodak any explanation on why the warning is there??

**Comment by rohan100jain** [comment]
I can trace the warning down to https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/module/module.py#L338 where we pass in a set() to nest.flatten which triggers the warning. I don't think this is something you need to really worry about. Assigning to Tom who wrote tf.Module

**Comment by kevinashaw** [comment]
I'm seeing the same error in TF2.2 in Python 3.7 on MacOS. 
```
2020-06-21 23:12:15.321826: W tensorflow/python/util/util.cc:329] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
WARNING:tensorflow:From /Users/.../anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/resource_variable_ops.py:1817: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.
Instructions for updating:
If using Keras pass *_constraint arguments to layers.

```
Error depends on this line.  The error dissappears when this line is commented.
`model.save(os.path.join(run_dir, f'model_{run_dirname}_keras_best'), save_format='tf')
`
I agree that the error is probably not harmful, but it is annoying.

**Comment by Saduf2019** [comment]
@songs18  
I ran the code on tf nightly and do not see nay warnings,please have a look at the [gist here](https://colab.research.google.com/gist/Saduf2019/99b833665239d84bc2568930506699ec/untitled268.ipynb).
Could you please confirm and let us know.

**Comment by songs18** [comment]
@Saduf2019 
This issue occurred at tensorflow  v2.0.0.
I see you installed tf-nightly-2.4.0.dev20200714 in the given gist.

**Comment by kevinashaw** [comment]
I am still seeing it on 2.3.0-rc0 and on a recent 2.4-nightly as well.
Here is the full error message from a run last night with 2.3.0-rc0.
Personally, I can't tell if this is one error or two that overlap.
```
WARNING:tensorflow:From /Users/kevinshaw/anaconda3/lib/python3.7/site-packages/tensorflow/python/training/tracking/tracking.py:111: Model.state_updates (from tensorflow.python.keras.engine.training) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
2020-07-12 22:57:22.189856: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
WARNING:tensorflow:From /Users/kevinshaw/anaconda3/lib/python3.7/site-packages/tensorflow/python/training/tracking/tracking.py:111: Layer.updates (from tensorflow.python.keras.engine.base_layer) is deprecated and will be removed in a future version.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.

```

**Comment by Saduf2019** [comment]
@songs18
I have verified on colab for 2.0 as well, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/85d4d135c772ddcdb7d01a0537dbcad2/untitled265.ipynb)

@kevinashaw 
Can you please let us know where are you seeing these warnings.

**Comment by kevinashaw** [comment]
@Saduf2019 
As best I can tell if occurs during the `model.fit()` operation.  
In particular, it occurs between the start of the operation and before the first `Epoch 1/90` statement.
Tensorflow graph messages are always cryptic since it is almost impossible to tell where in the graph the error occurred.

**Comment by songs18** [comment]
@Saduf2019 , @kevinashaw 
Thanks.
After a few months, I reviewed the code again and found the problem. 
This warning occurs when the shape of the model output does not match with the ground_truth labels. Although it can be broadcasted.
However, the warning given is obscure and it can be better described.

**Comment by kevinashaw** [comment]
Interestingly, I think this has to do with my use of `tf.squeeze()` at the end of my model.  I wanted to remove the channels-dimension after some 1x1 CNNs, such that `(None, 155, 1)` would be reduced to `(None, 155)` and I choose (reasonably) to use `tf.squeeze()` however it apparently returns a shape of `[None]`.  While the model would still train, this caused much confusion and i think this was forcing TF to broadcast into the target vector, since it did not know the true shape of the output tensor.

```
__________________________________________________________________________________________________
L08_1x1CNN (Conv1D)             (None, 155, 1)       601         L07_Dropout[0][0]                
__________________________________________________________________________________________________
tf_op_layer_L09_Squeeze (Tensor [None]               0           L08_1x1CNN[0][0]                 
__________________________________________________________________________________________________
tf_op_layer_QrsTarget (TensorFl [None]               0           tf_op_layer_L09_Squeeze[0][0]    
==================================================================================================
```

**Comment by levidehaan** [comment]
Does anyone else think it's weird that @Saduf2019 is on here asking people to post their entire codebase, has no code or repos in their repo, etc..? Is there any way to block these kinds of scammers from being able to post on these kinds of repos. Luckily the OP is smart enough to know that this person was trying to get their codebase, but others may not be so wary and post things that are not necessary to fix issues.
